crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -880,6 +881,34 @@ where
 
         self.current_match_case.as_mut().unwrap().index += 1;
     }
+
+    fn visit_except_handler(&mut self, except_handler: &'ast ast::ExceptHandler) {
+        let ast::ExceptHandler::ExceptHandler(ast::ExceptHandlerExceptHandler {
+            name,
+            type_,
+            body,
+            range: _,
+        }) = except_handler;
+        if let Some(type_) = type_ {
+            if let Some(symbol_name) = name {
+                self.add_standalone_expression(type_);
```

Comment:
Is this the change you're suggesting for this function?

```diff
diff --git a/crates/red_knot_python_semantic/src/semantic_index/builder.rs b/crates/red_knot_python_semantic/src/semantic_index/builder.rs
index 80210a11c..668ad4b77 100644
--- a/crates/red_knot_python_semantic/src/semantic_index/builder.rs
+++ b/crates/red_knot_python_semantic/src/semantic_index/builder.rs
@@ -889,22 +889,18 @@ where
             body,
             range: _,
         }) = except_handler;
-        if let Some(type_) = type_ {
+        if let Some(handled_exceptions) = type_ {
+            self.visit_expr(handled_exceptions);
             if let Some(symbol_name) = name {
-                self.add_standalone_expression(type_);
-                self.visit_expr(type_);
-                debug_assert!(self.current_assignment.is_none());
                 let symbol =
                     self.add_or_update_symbol(symbol_name.id.clone(), SymbolFlags::IS_DEFINED);
                 self.add_definition(
                     symbol,
                     ExceptHandlerDefinitionNodeRef {
                         symbol_name,
-                        handled_exceptions: type_,
+                        handled_exceptions,
                     },
                 );
-            } else {
-                self.visit_expr(type_);
             }
         }
         self.visit_body(body);
```

It causes the tests I added to panic:

```
failures:

---- types::infer::tests::except_handler stdout ----
thread 'types::infer::tests::except_handler' panicked at crates/red_knot_python_semantic/src/semantic_index.rs:207:33:
no entry found for key

---- types::infer::tests::invalid_except_handler stdout ----
thread 'types::infer::tests::invalid_except_handler' panicked at crates/red_knot_python_semantic/src/semantic_index.rs:207:33:
no entry found for key

---- types::infer::tests::unknown_type_in_except_handler_does_not_cause_spurious_diagnostic stdout ----
thread 'types::infer::tests::unknown_type_in_except_handler_does_not_cause_spurious_diagnostic' panicked at crates/red_knot_python_semantic/src/semantic_index.rs:207:33:
no entry found for key


failures:
    types::infer::tests::except_handler
    types::infer::tests::invalid_except_handler
    types::infer::tests::unknown_type_in_except_handler_does_not_cause_spurious_diagnostic

test result: FAILED. 236 passed; 3 failed; 2 ignored; 0 measured; 0 filtered out; finished in 0.18s
```

Reply:
Did you also change the code I mentioned in `infer.rs` that currently calls `infer_expression_types` salsa query on this expression. That code requires the `add_standalone_expression` call here -- but we don't need either one.

Reply:
I don't like that this results in a panic instead of somehow being enforced by the type system; I would rather that we can use a salsa query on any expression. But that's a necessary concession to salsa performance issues. And since our base representation is the AST, which we can't mutate, we are left with this fundamental issue that some `ast::Expr` have a matching `Expression` tracked structnwe can query on and some do not, and the type system doesn't know the difference.

Reply:
> Did you also change the code I mentioned in `infer.rs` that currently calls `infer_expression_types` salsa query on this expression, but should instead just use `self.infer_expression`? That call to `infer_expression_types` query requires this call to `add_standalone_expression` here -- but we don't need either one.

Yup, making that change at the same time fixed it. Thanks :-)